### PR TITLE
8288298: Resolve multiline message parsing ambiguities in UL

### PR DESCRIPTION
**Pull request by anton-seoane**
Tools such as JITWatch parse OpenJDK logs (e.g. logs generated from `LogCompilation`) in order to present interesting data to users. In order for these tools to work reliably, UnifiedLogging (UL) needs to have a consistent output scheme.

Currently, a UL log message can be formatted in such a way that it looks like a UL decorator prefix, causing issues with parsing. This is because logging functions in UL do not prepend decorators (or any kind of prefix) to newlines in a log message. For example, `log_info(gc)("A\nB");` currently outputs
```
[0s][gc] A
B
```
and we could mistakenly interpret `B` as a decorator. Additionally, developers may introduce pseudo-decorators (something that looks like a decorator but is actually part of the log message), yielding an incorrect parse. As a side remark, but also relevant, we hinder human readability when logs appear suddenly skewed at some points.

The UL framework should offer a robust way to (a) distinguish decorators from messages, and (b) unambiguously group multiline output. This PR aims to achieve both goals through a subset of the changes proposed in [this mail](https://mail.openjdk.org/pipermail/hotspot-dev/2024-October/095284.html), from which the rest of the ideas can be done as future work. This means:

- Decorators cannot contain the symbols `[` or `]`. The special decorator `[ ]` (with a variable number of whitespace between the brackets) is reserved.
- We separate decorators from log messages via the first space after a closing bracket in the line.
- A log message can contain any kind of symbols, and ends with a newline (except in the case for multiline messages).
- We prepend multiline logging (such as in the example above) with the invalid decorator `[ ]`. The invalid decorator is as wide as the indentation of the rest of the log for easy visual reading. For example:
```
[0s][gc] single-line message
[1s][gc] another single line message
[2s][gc] first line of a multiline message
[      ] second and last line of a multiline message
[3s][gc] another single line message
```
Note how this is both unambiguously parseable and human readable.

For the case where decorators have been disabled, the aforementioned points do not apply (i.e., behaviour remains the same as before). This means no multiline logical connection (such as the one presented above) and no way to separate decorators (the empty set, in this case) from messages. This is intentional as users specifying no decorators expect "raw" output. Additionally, it is to be assumed that the end user is in control of the log command (and effectively disabling decorators means that they are aware of no multiline grouping and no unambiguous parsing solution).

If the option `-Xlog:foldmultilines` (already present in UL) is specified, we do not carry out any multiline grouping.

Performance degradation is negligible (<1%) for common `-Xlog` usages up to trace level. Extreme test case `-Xlog:all=trace` only yields an 8% performance overhead.

---------
### Progress
- [ ] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8288298](https://bugs.openjdk.org/browse/JDK-8288298): Resolve multiline message parsing ambiguities in UL (**Enhancement** - P4)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/21501/head:pull/21501` \
`$ git checkout pull/21501`

Update a local copy of the PR: \
`$ git checkout pull/21501` \
`$ git pull https://git.openjdk.org/jdk.git pull/21501/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 21501`

View PR using the GUI difftool: \
`$ git pr show -t 21501`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/21501.diff">https://git.openjdk.org/jdk/pull/21501.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/21501#issuecomment-2434731617)

### CODE REVIEW

File: src/hotspot/share/logging/logFileStreamOutput.cpp [hunk 1]
```diff
@@ -146,6 +164,28 @@ int LogFileStreamOutput::write_internal(const LogDecorations& decorations, const
   return written;
 }
 
+int LogFileStreamOutput::write_internal(const LogDecorations& decorations, const char* msg) {
+  int msg_len = (int)strlen(msg);
```
Comment (jdksjolen) [inline src/hotspot/share/logging/logFileStreamOutput.cpp:168]: Not OK with casting this. Add some sort of overflow check instead and only then `static_cast`.

File: src/hotspot/share/logging/logFileStreamOutput.cpp [hunk 2]
```diff
@@ -146,6 +164,28 @@ int LogFileStreamOutput::write_internal(const LogDecorations& decorations, const
...
+
+  // Do not do anything if foldmultilines has been specified
```
Comment (jdksjolen) [inline src/hotspot/share/logging/logFileStreamOutput.cpp:170]: But we do something, right? Maybe "don't alter the string if _fold_multilines has been specified". Please add a corresponding explanation to the other case, probably something mentioning "canonical representation" or something of the sort.

File: src/hotspot/share/logging/logFileStreamOutput.cpp [hunk 3]
```diff
@@ -146,6 +164,28 @@ int LogFileStreamOutput::write_internal(const LogDecorations& decorations, const
...
+    if (*tmp == '\n') {
+      *tmp = '\0';
+    }
+    ++tmp;
+  }
+  int written = write_internal_line(decorations, dupstr, msg_len);
+
+  os::free(dupstr);
+
+  return written;
```
Comment (jdksjolen) [inline src/hotspot/share/logging/logFileStreamOutput.cpp:186]: I don't think we need to dupstr this, do we? Can't we do the changes in the original string instead?

**Review by jdksjolen — COMMENTED** [review]
The proposal itself seems OK. Do we have a regex or something to parse this? I need to think about its unambiguity a bit closer.

It would technically be OK for decorators to have spaces inside of them (and even newlines?) as long as it's not *only* spaces or newlines, right? Anyway, it doesn't particularly matter, the restrictions as set out are fine.

A few code questions, I'll get back to you later on.

File: src/hotspot/share/logging/logFileStreamOutput.cpp (hunk 3 above)
Comment (anton-seoane) [reply src/hotspot/share/logging/logFileStreamOutput.cpp:186]: We would lose the `const`ness of the original string. Intuitively I'm not sure either if a printing function should modify the string passed to it.

In any way, the effect of this PR in performance is really negligible, so "duplicating or not" would be a case of neatness more than optimisation. I'm not sure about what's neater, though, so I'll return the question to you

Comment (jdksjolen) [reply src/hotspot/share/logging/logFileStreamOutput.cpp:188]: We'll leave it for future work. It can definitely be done, but is not necessary yet.

**Review by jdksjolen — COMMENTED** [review]
The performance degradation is caused by three things:

1. Walking the string with `strlen`
2. Copying the string with `strdup` (or `memcopy` soon)
3. Walk the string to switch out newlines with nul bytes

I happen to know that both steps 1 and 2 are unnecessary, and that re-architecting UL slightly let's us avoid these steps here entirely. We already know the length up the call chain and these strings are actually not `const`.

The third step is necessary, but is also amenable to vectorization.

**Review by jdksjolen — COMMENTED** [review]
Some more comments.

**Pull request closed without merging**